core/src/main/scala/com/avast/grpc/jsonbridge/ReflectionGrpcJsonBridge.scala

```diff
@@ -0,0 +1,191 @@
+              case Right(req) => execute(req, headers).map(resp => Right(printer.print(resp)))
+              case Left(status) => F.pure(Left(status))
+            }
+          }
+
+          (method.getMethodDescriptor.getFullMethodName, handler)
+        }
+    }.toMap
+
+  override def invoke(methodName: GrpcMethodName, body: String, headers: Map[String, String]): F[Either[Status, String]] =
+    handlersPerMethod.get(methodName.fullName) match {
+      case None => F.pure(Left(Status.NOT_FOUND.withDescription(s"Method '$methodName' not found")))
+      case Some(handler) =>
+        handler(body, headers)
+          .recover {
+            case NonFatal(ex) =>
+              val message = "Error while executing the request"
+              logger.info(message, ex)
+              ex match {
+                case e: StatusException if e.getStatus.getCode == Status.Code.UNKNOWN =>
+                  Left(richStatus(Status.INTERNAL, message, e.getStatus.getCause))
+                case e: StatusRuntimeException if e.getStatus.getCode == Status.Code.UNKNOWN =>
+                  Left(richStatus(Status.INTERNAL, message, e.getStatus.getCause))
+                case e: StatusException =>
+                  Left(richStatus(e.getStatus, message, e.getStatus.getCause))
+                case e: StatusRuntimeException =>
+                  Left(richStatus(e.getStatus, message, e.getStatus.getCause))
+                case _ =>
+                  Left(richStatus(Status.INTERNAL, message, ex))
+              }
+          }
+    }
+
+  override val methodsNames: Seq[GrpcMethodName] = handlersPerMethod.keys.map(m => GrpcMethodName(m)).toSeq
+  override val servicesNames: Seq[String] = methodsNames.map(_.service).distinct
+
+}
+
+object ReflectionGrpcJsonBridge extends StrictLogging {
+
+  private val parser: JsonFormat.Parser = JsonFormat.parser()
+
+  private val printer: JsonFormat.Printer = {
+    JsonFormat.printer().includingDefaultValueFields().omittingInsignificantWhitespace()
+  }
+
+  private def createNewFutureStubFunction[F[_]](inProcessChannel: Channel, ssd: ServerServiceDefinition)(
+      implicit F: Sync[F]): F[AbstractStub[_]] = F.delay {
+    val method = getServiceGeneratedClass(ssd.getServiceDescriptor).getDeclaredMethod("newFutureStub", classOf[Channel])
+    method.invoke(null, inProcessChannel).asInstanceOf[AbstractStub[_]]
+  }
+
+  private def executeRequest[F[_]](createFutureStub: F[AbstractStub[_]], requestMessagePrototype: Message, javaMethodName: String)(
+      req: Message,
+      headers: Map[String, String])(implicit executor: Executor, F: Async[F]): F[MessageOrBuilder] = {
+
+    def createMetadata(): Metadata = {
+      val md = new Metadata()
+      headers.foreach { case (k, v) => md.put(Metadata.Key.of(k, Metadata.ASCII_STRING_MARSHALLER), v) }
+      md
+    }
+
+    for {
+      futureStub <- createFutureStub
+      method = futureStub.getClass.getDeclaredMethod(javaMethodName, requestMessagePrototype.getClass)
```

Comment:
@jendakol Please move this line to a place where it is called just once - this call has significant performance impact. Also subsequent `invoke` call on _always-new_ instance is not so effective.